File: Stripe/STPPaymentContextAmountModel.m
//
//  STPPaymentContextAmountModel.m
//  Stripe
//
//  Created by Brian Dorfman on 8/16/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

#import "STPPaymentContextAmountModel.h"

#import "NSDecimalNumber+Stripe_Currency.h"

@implementation STPPaymentContextAmountModel {
    NSInteger _paymentAmount;
    NSArray<PKPaymentSummaryItem *> *_paymentSummaryItems;
}

- (instancetype)initWithAmount:(NSInteger)paymentAmount {
    self = [super init];
    if (self) {
        _paymentAmount = paymentAmount;
        _paymentSummaryItems = nil;
    }
    return self;
}

- (instancetype)initWithPaymentSummaryItems:(NSArray<PKPaymentSummaryItem *> *)paymentSummaryItems {
    self = [super init];
    if (self) {
        _paymentAmount = 0;
        _paymentSummaryItems = paymentSummaryItems;
    }
    return self;
}

- (NSInteger)paymentAmountWithCurrency:(NSString *)currency shippingMethod:(PKShippingMethod *)shippingMethod {
    NSInteger shippingAmount = (shippingMethod != nil) ? [shippingMethod.amount stp_amountWithCurrency:currency] : 0;
    if (_paymentSummaryItems == nil) {
        return _paymentAmount + shippingAmount;
    }
    else {
        PKPaymentSummaryItem *lastItem = _paymentSummaryItems.lastObject;
        return [lastItem.amount stp_amountWithCurrency:currency] + shippingAmount;
    }
}

- (NSArray<PKPaymentSummaryItem *> *)paymentSummaryItemsWithCurrency:(NSString *)currency
                                                         companyName:(NSString *)companyName
                                                      shippingMethod:(PKShippingMethod *)shippingMethod {
    PKPaymentSummaryItem *shippingItem = nil;
    if (shippingMethod != nil) {
        shippingItem = [PKPaymentSummaryItem summaryItemWithLabel:shippingMethod.label
                                                           amount:shippingMethod.amount];
    }
    if (_paymentSummaryItems == nil) {
        NSInteger shippingAmount = [shippingMethod.amount stp_amountWithCurrency:currency];
        NSDecimalNumber *total = [NSDecimalNumber stp_decimalNumberWithAmount:_paymentAmount + shippingAmount
                                                                     currency:currency];
        PKPaymentSummaryItem *totalItem = [PKPaymentSummaryItem summaryItemWithLabel:companyName
                                                                              amount:total];
        NSMutableArray *items = [@[totalItem] mutableCopy];
        if (shippingItem != nil) {
            [items insertObject:shippingItem atIndex:0];
        }
        return [items copy];
    }
    else {
        if ([_paymentSummaryItems count] > 0 && shippingItem != nil) {
            NSMutableArray *items = [_paymentSummaryItems mutableCopy];
            PKPaymentSummaryItem *origTotalItem = [items lastObject];
            NSDecimalNumber *newTotal = [origTotalItem.amount decimalNumberByAdding:shippingItem.amount];
            PKPaymentSummaryItem *totalItem = [PKPaymentSummaryItem summaryItemWithLabel:origTotalItem.label amount:newTotal];
            [items removeLastObject];
            return [[items arrayByAddingObjectsFromArray:@[shippingItem, totalItem]] copy];
        }
        else {
            return _paymentSummaryItems;
        }
    }
}

@end

File: Stripe/STPPaymentContextAmountModel.swift
//
//  STPPaymentContextAmountModel.swift
//  Stripe
//
//  Created by Brian Dorfman on 8/16/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

import Foundation
import PassKit

/// Internal model for STPPaymentContext's `paymentAmount` and
/// `paymentSummaryItems` properties.
class STPPaymentContextAmountModel: NSObject {
  private var paymentAmount = 0
  private var paymentSummaryItems: [PKPaymentSummaryItem]?

  init(amount paymentAmount: Int) {
    super.init()
    self.paymentAmount = paymentAmount
    paymentSummaryItems = nil
  }

  init(paymentSummaryItems: [PKPaymentSummaryItem]?) {
    super.init()
    paymentAmount = 0
    self.paymentSummaryItems = paymentSummaryItems
  }

  func paymentAmount(withCurrency currency: String?, shippingMethod: PKShippingMethod?) -> Int {
    let shippingAmount =
      ((shippingMethod != nil) ? shippingMethod?.amount.stp_amount(withCurrency: currency) : 0) ?? 0
    if paymentSummaryItems == nil {
      return paymentAmount + shippingAmount
    } else {
      let lastItem = paymentSummaryItems?.last
      return (lastItem?.amount.stp_amount(withCurrency: currency) ?? 0) + shippingAmount
    }
  }

  func paymentSummaryItems(
    withCurrency currency: String?,
    companyName: String?,
    shippingMethod: PKShippingMethod?
  ) -> [PKPaymentSummaryItem]? {
    var shippingItem: PKPaymentSummaryItem?
    if let shippingMethod = shippingMethod {
      shippingItem = PKPaymentSummaryItem(
        label: shippingMethod.label,
        amount: shippingMethod.amount)
    }
    if paymentSummaryItems == nil {
      let shippingAmount = shippingMethod?.amount.stp_amount(withCurrency: currency) ?? 0
      let total = NSDecimalNumber.stp_decimalNumber(
        withAmount: paymentAmount + shippingAmount,
        currency: currency)
      var totalItem: PKPaymentSummaryItem?
      if let total = total {
        totalItem = PKPaymentSummaryItem(
          label: companyName ?? "",
          amount: total)
      }
      var items = [totalItem]
      if let shippingItem = shippingItem {
        items.insert(shippingItem, at: 0)
      }
      return items.compactMap { $0 }
    } else {
      if (paymentSummaryItems?.count ?? 0) > 0 && shippingItem != nil {
        var items = paymentSummaryItems
        let origTotalItem = items?.last
        var newTotal: NSDecimalNumber?
        if let amount1 = shippingItem?.amount {
          newTotal = origTotalItem?.amount.adding(amount1)
        }
        var totalItem: PKPaymentSummaryItem?
        if let newTotal = newTotal {
          totalItem = PKPaymentSummaryItem(label: origTotalItem?.label ?? "", amount: newTotal)
        }
        items?.removeLast()
        if let items = items {
          return items + [shippingItem, totalItem].compactMap { $0 }
        }
        return nil
      } else {
        return paymentSummaryItems
      }
    }
  }
}
